go/apps/mixnet/README.md

```diff
@@ -0,0 +1,168 @@
+particular protocol. However, this would require the recipients to participate
+in the protocol, since messages exiting the mixnet must be encrypted.
+
+For a protocol to achieve security in this sense, the messages must all have the
+same length; of course, this is not always reasonable in practice. Mixnets
+address this by splitting messages into fixed-length cells. Senders split
+messages into cells (zero-padding the last cell as needed) and send them to the
+first router where they are added to a queue. At each round the router waits
+until there are _m_ cells in the queue from _m_ distinct senders and transmits
+these simultaneously to achieve anonymity.
+
+Extending the definition of unlinkability to a mixnet that divides messages into
+cells and transmits at rounds is challenging: the presence of
+variable-length messages exposes traffic patterns to the adversary. One way to
+mitigate this problem is to zero-pad all messages to the length of the longest
+message. This achieves a property called _unobservability_ [1] which is too
+expensive for our purposes.
+Another appraoch is to weaken the security model to one in which the adversary
+may only observe a fraction of the network at any one time; relaying messages
+over circuits of routers may make it more difficult to perform traffic
+analysis. This is the case for the design of the Tor onion-routing protocol [2].
+We will consider extending our protocol to a network of routers to achieve
+security in this model. This approach may have the added benifit of reducing
+latency of messages traversing the mixnet.
+
+Design
+------
+
+We designed the mixnet to proxy client/server protocols. There are
+two main components of the protocol: the _proxy_ (`mixnet_proxy`) accepts
+arbitrary TCP connections from a client and relays messages over the
+mixnet to the server; the _router_ (`mixnet_router`) accepts connections
+from a proxy and performs the mixnet operations. The proxy and router perform
+a one-way authenticated TLS handshake to exchange a key for wrapping
+(encrypting and authenticating) messages sent between them. Messages sent
+between proxies and routers (or routers and routers) are fixed length cells.
+
+The protocol for a single proxy and router is as follows.
+To send a message to a server, the client sends the message to the proxy
+which divides the message into cells and sends them to the router. The
+router waits until it has received all the cells, then assembles them into
+the original message. Once the router has queued messages from enough senders, it
+transmits the message to the server. It then waits for a reply from the server,
+divides the reply into cells, and queues the cells to be transmitted back to
+the proxy. The proxy waits until it has received all the reply cells, assembles
+them into the complete message, and sends it back to the client.
+
+The router maintains two datastructures: the _sendQueue_ for sender to recipient
```

Comment:
"datastructures" -> "data structures". I don't think this noun phrase has become a single word yet, unlike the words "database" and "filesystem".